### PR TITLE
WIP: Add Validation Section

### PR DESCRIPTION
**Pull request by mmccool**
Add a validation subsection to the end of the Serialization section.  Defines three levels of validation appropriate to different contexts.

* Resolves Issue https://github.com/w3c/wot-thing-description/issues/1043
* Relevant to Issue https://github.com/w3c/wot-discovery/issues/143 (mostly resolves it, although we would still need to add a reference from WoT Discovery to Intermediate Validation for directories that do semantic processing; to discuss, how to deal with extensions...)
* Relevant to Issue https://github.com/w3c/wot-discovery/issues/99 (mostly resolves it, although we still need to add references to WoT Discovery to cite the appropriate levels of validation).

NOTES: 
* also added missing id in immediately preceeding Identification section.
* adjacent to proposed canonicalization section so order will depend on which one is merged first

- [x] define Basic Validation
- [ ] discuss and refine Basic validation definition
- [x] define Intermediate Validation
- [ ] discuss and refine Intermediate validation definition
- [x] define Full Validation
- [ ] discuss and refine Full validation definition
- [ ] add correct section links
- [ ] reference JSON schema in appendix
- [ ] add special case for TDs with no extensions (may need additional assertions, i.e. not allowing extra keywords if not defined in an extension)
- [ ] define/reference how context and SHACL files are to be fetched over the network
- [ ] provide SHACL files for TD
- [ ] label appropriate assertions with Basic tag
- [ ] label appropriate assertions with Intermediate tag
- [ ] Document tags in RFC2119 introduction section
- [ ] update change log

***
<a href="https://pr-preview.s3.amazonaws.com/mmccool/wot-thing-description/pull/1085.html" title="Last updated on May 5, 2021, 3:37 PM UTC (f4f38cf)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/wot-thing-description/1085/0e54c8f...mmccool:f4f38cf.html" title="Last updated on May 5, 2021, 3:37 PM UTC (f4f38cf)">Diff</a>

### CODE REVIEW

**Comment by egekorkan** [comment]
Looks good so far but is this ready for review @mmccool ? 

One point that is not clear, would assertions that are **not** tagged with basic or intermediate always fall into full validation category? It might be better to also add tags containing `full`

**Comment by mmccool** [comment]
If marked WIP or draft than not ready for review.

**Comment by mmccool** [comment]
Anyhow, the idea of "not tagged" being full is that this way we can't possibly miss one.  The only question is, are there any assertions that can't be validated, even in principle?  Or that are sort of meta, like "preserve the privacy of users".  I can see omitting things like that from "full validation", simply because they can only be asserted by implementers (making a promise not to be evil) and cannot be tested by an automated system.

**Comment by mmccool** [comment]
BTW we discussed this in the plugfest call today and our plan is to update the implementation report generation scripts and the validation playground in the next couple of weeks.  One output of that will be to label which assertions can be automatically tested at the "Basic" level.

Also for Directories I will probably point at the Intermediate level.

**Comment by sebastiankb** [comment]
we will also introduce a new namespace for TM. Also see here https://github.com/w3c/wot-thing-description/pull/1092#issuecomment-819628112

It would be great to cover this also in the validation.

File: index.html [hunk 1]
```diff
@@ -1450,10 +1450,10 @@ <h2>Data Schema Vocabulary Definitions</h2>
...
                        information in different languages. Also see <cite><a href='#multilanguage'>MultiLanguage</a></cite>.</td><td>optional</td><td><a>Map</a> of <a href="http://www.w3.org/TR/2012/REC-xmlschema11-2-20120405/#string"><code>string</code></a></td></tr>
-<tr class="rfc2119-table-assertion" id="td-vocab-const--DataSchema"><td><code>const</code></td><td>Provides a constant value.</td><td>optional</td><td>any type</td></tr>
 <tr class="rfc2119-table-assertion" id="td-vocab-unit--DataSchema"><td><code>unit</code></td><td>Provides unit information that is used, e.g.,
                 in international science, engineering, and
                 business.</td><td>optional</td><td><a href="http://www.w3.org/TR/2012/REC-xmlschema11-2-20120405/#string"><code>string</code></a> or <a>Array</a> of <a href="http://www.w3.org/TR/2012/REC-xmlschema11-2-20120405/#string"><code>string</code></a></td></tr>
+<tr class="rfc2119-table-assertion" id="td-vocab-const--DataSchema"><td><code>const</code></td><td>Provides a constant value.</td><td>optional</td><td>any type</td></tr>
```
Comment (relu91) [inline index.html:1456]: I've seen this term popping up and be removed in different PRs... not sure if there were a rendering problem somewhere in the commit history.

**Comment by sebastiankb** [comment]
Regarding namespace for TM, also see the new PR #1123

File: index.template.html [hunk 2]
```diff
@@ -2992,6 +2992,73 @@ <h2>Identification</h2>
...
+  validated given only the JSON serialization, for instance, the assertions
+  listed under <a>Behavioral Assertions</a> that relate a TD to the
+  behavior of a Thing that it describes.  Extensions are also problematic,
+  in that even if formal metadata is given for validating an extension,
+  dynamically fetching this metadata in a deployment might pose a privacy
+  risk.  In this section, therefore, we name and define various levels of
+  validation appropriate for different contexts.
+  </p>
+  <section id="basic-validation-serialization-json">
+  <h3>Basic Validation</h3>
```
Comment (sebastiankb) [inline index.template.html:3013]: Maybe we can also call this as "Minimum Validation"

File: index.template.html [hunk 3]
```diff
@@ -2992,6 +2992,73 @@ <h2>Identification</h2>
...
+  in combination with a few spot checks, for example to check that 
+  security schema names have matching definitions.
+  </p><p> 
+  This level of validation includes all assertions implied by 
+  normative tables in this document as well as all assertions marked
+  with the "Basic" tag.
+  </p>
+  </section>
+  <section id="intermediate-validation-serialization-json">
+  <h3>Intermediate Validation</h3>
```
Comment (sebastiankb) [inline index.template.html:3028]: I'm not sure if "Intermediate" is the best naming. If we use "Minimum Validation" above, maybe we can use "Basic Validation" here?

File: index.template.html [hunk 4]
```diff
@@ -2992,6 +2992,73 @@ <h2>Identification</h2>
...
+  context definitions and SHACL constraints for extension vocabularies 
+  can be fetched, then these
+  can be used to validate extensions.
+  </p><p> 
+  This level of validation includes all those covered by <a>Basic Validation</a>
+  as well as all assertions marked with the "Intermediate" tag.
+  </p>
+  </section>
+  <section id="full-validation-serialization-json">
+  <h3>Full Validation</h3>
```
Comment (sebastiankb) [inline index.template.html:3047]: maybe "Advance Validation"

**Comment by sebastiankb** [comment]
I agree to merge this draft to have this in the next WD version. The next update should include links / references to JSON schema & SHACL definitions for TD & TM. I have some alternative ideas on the different validation levels, but this can also be changed in the next update.

**Comment by sebastiankb** [comment]
from today's TD call we decided to merge
